### PR TITLE
implement astgen union + enum & update behavior tests

### PR DESCRIPTION
**Pull request by Rexicon226**
i believe this finishes up #17916. 

might have missed some cases, please let me know!

### CODE REVIEW

File: src/Sema.zig
Comment (mlugg) [inline src/Sema.zig:20866]: This logic is for reification (`@Type`) - it should not be deleted.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:5305]: No point deiniting these since they're allocated into the arena+SFBA. Same for struct decls - sorry for missing that in the last PR.

**Comment by Rexicon226** [comment]
@mlugg sorry to bother :D whenever you have a chance to teview

**Review by Vexu — APPROVED** [review]

File: test/cases/compile_errors/duplicate_union_field.zig
Comment (nektro) [inline test/cases/compile_errors/duplicate_union_field.zig:15]: I don't like that it's reversed the order of which one shows up as the duplicate

Comment (Vexu) [reply test/cases/compile_errors/duplicate_union_field.zig:15]: The order might be reversed but it makes more sense to list the duplicates top to bottom now that there can be multiple.

Comment (nektro) [reply test/cases/compile_errors/duplicate_union_field.zig:15]: maybe `union field name conflict: 'Bar'` ?

Comment (Rexicon226) [reply test/cases/compile_errors/duplicate_union_field.zig:15]: something like:
```
:2:5: error: union field name conflict: 'Bar'
:3:5: note: duplicate field here
```

Comment (Rexicon226) [reply test/cases/compile_errors/duplicate_union_field.zig:15]: after some thought i agree with this... the current error messages are a backwards. 

ill change the order around 👍

File: src/AstGen.zig
Comment (nektro) [inline src/AstGen.zig:1766]: field name is unquoted here

File: test/cases/compile_errors/duplicate_field_in_discarded_anon_init.zig
Comment (nektro) [inline test/cases/compile_errors/duplicate_field_in_discarded_anon_init.zig:11]: materializes here

File: test/cases/compile_errors/duplicate_field_in_struct_value_expression.zig
Comment (nektro) [inline test/cases/compile_errors/duplicate_field_in_struct_value_expression.zig:22]: materializes here

File: test/cbe.zig
Comment (nektro) [inline test/cbe.zig:512]: materializes here

**Review by nektro — CHANGES_REQUESTED** [review]

**Comment by Rexicon226** [comment]
will merge the latest changes from master to re-run the ci, maybe itll fix the flaky test 🤔

**Comment by matu3ba** [comment]
> will merge the latest changes from master to re-run the ci, maybe itll fix the flaky test 🤔

Try
```diff
diff --git a/lib/std/net/test.zig b/lib/std/net/test.zig
index caac8f4e6..69f40aba9 100644
--- a/lib/std/net/test.zig
+++ b/lib/std/net/test.zig
@@ -356,6 +356,7 @@ test "non-blocking tcp server" {
     defer socket_file.close();
 
     var client = try server.accept();
+    defer client.stream.close();
     const stream = client.stream.writer();
     try stream.print("hello from server\n", .{});
```

**Comment by Rexicon226** [comment]
saw the discussion earlier about this 😃 

the ci has already passed when i rebased so it wouldn't make much sense now, but i hope it helps in the future.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:1765]: Assuming this wasn't specifically requested by a core team member for any reason: you really don't need to put the field name here. This information already exists: it's the token the error points to. Adding more words to the error message just makes it take longer to read.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:1792]: The deletion of this comment indicates that this logic is no longer necessary. Either rework the logic (to evaluate each field individually with the `discard` result location rather than constructing any aggregate), or put a new TODO comment here.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:5080]: Relating to first comment: okay, fair enough, this was inconsistent between inits vs type decls before. I think the best option is to *not* have the name since, as discussed above, it's redundant information. Also, `duplicate struct field` is a much clearer message than `struct field name conflict`; I'm not sure why you felt the need to change that bit.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:5286]: Same as above.

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:5583]: And here.

**Review by mlugg — CHANGES_REQUESTED** [review]
My apologies for how long it took me to look over this - I've had a lot going on in my personal life over the past few weeks.

File: src/AstGen.zig
Comment (Rexicon226) [reply src/AstGen.zig:1765]: You're all good with the personal life stuff. I hope it went well (i believe you were moving?).

It wasn't requested by a core team member, and I agree, it is duplicate information. I'll remove it in the following commit, as well as update the tests.

File: src/AstGen.zig
Comment (Rexicon226) [reply src/AstGen.zig:1792]: I don't want to put another TODO :smile:
Please explain a little more exactly what you want me to do.

As I currently understand, this is no longer needed?

Comment (mlugg) [reply src/AstGen.zig:1792]: This corresponds to code like `_ = .{ .a = 0, .b = 1 }`. Rather than actually initing a struct temporary, this should just evaluate `0` and `1` with `ResultInfo.Loc.discard`, as if you wrote `_ = 0; _ = 1;`.

Comment (Rexicon226) [reply src/AstGen.zig:1792]: Are you saying that this type of discarded struct should not be checked for duplicates?

Comment (mlugg) [reply src/AstGen.zig:1792]: No, that is absolutely not what I am saying. As the removed comment mentions, the current logic - actually constructing a struct temporary - exists _in order to_ ensure this error is triggered. Since the logic is now moved to AstGen, it is not necessarily to construct a temporary anymore.

Comment (Rexicon226) [reply src/AstGen.zig:1792]: ok, thanks for clarifying. that's what i thought in my first message, but i began to doubt my understanding of english :smile:

i've looked through the code, and thought about what you said, and i understand now!

will just this be fine: `.discard => return .void_value,`

Comment (mlugg) [reply src/AstGen.zig:1792]: No - as I have said, it is still important to evaluate the sub-expressions, for any compilation errors or side effects. Before returning `.void_value`, call `expr` on each initialization sub-expression with `.{ .rl = .discard }` as the `ri`.

**Comment by andrewrk** [comment]
Needs a rebase and then another review from @mlugg

**Comment by Rexicon226** [comment]
> Needs a rebase and then another review from @mlugg

Can do right now!

File: src/AstGen.zig
Comment (mlugg) [inline src/AstGen.zig:5311]: Inconsistent error message: `duplicate struct field` vs `duplicate union field name`

File: src/Sema.zig
Comment (mlugg) [inline src/Sema.zig:36873]: Does the removal of this not make `field_name_table` unnecessary in the first place?

**Review by mlugg — CHANGES_REQUESTED** [review]

File: src/AstGen.zig
Comment (Rexicon226) [reply src/AstGen.zig:5311]: done 👍

File: src/Sema.zig
Comment (Rexicon226) [reply src/Sema.zig:36873]: done 👍

**Comment by mlugg** [comment]
Nice work, thanks!

**Pull request merged**
